internal/pkg/source/source_test.go

```diff
@@ -0,0 +1,166 @@
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"testing"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/analysistest"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/ssa"
+)
+
+type testConfig struct {
+	propagatorsPattern string
+	fieldsPattern      string
+	sanitizerPattern   string
+}
+
+func (c *testConfig) IsSanitizer(call *ssa.Call) bool {
+	match, _ := regexp.MatchString(c.sanitizerPattern, call.String())
+	return match
+}
+
+func (c *testConfig) IsPropagator(call *ssa.Call) bool {
+	match, _ := regexp.MatchString(c.propagatorsPattern, call.String())
+	return match
+}
+
+func (c *testConfig) IsSourceFieldAddr(field *ssa.FieldAddr) bool {
+	match, _ := regexp.MatchString(c.fieldsPattern, field.String())
+	return match
+}
+
+type analyzerResult struct {
+	allocations []*ssa.Alloc
+	calls       []*ssa.Call
+	fieldAddr   []*ssa.FieldAddr
+	store       []*ssa.Store
+}
+
+var testAnalyzer = &analysis.Analyzer{
+	Name:       "source",
+	Run:        run,
+	Doc:        "test harness for the logic related to sources",
+	Requires:   []*analysis.Analyzer{buildssa.Analyzer},
+	ResultType: reflect.TypeOf(analyzerResult{}),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	in := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	var result analyzerResult
+	for _, fn := range in.SrcFuncs {
+		fmt.Printf("Func: %v\n", fn.Name())
+		for _, b := range fn.Blocks {
+			for _, i := range b.Instrs {
+				fmt.Printf("%v : %T\n", i, i)
+
+				switch v := i.(type) {
+				case *ssa.Alloc:
+					result.allocations = append(result.allocations, v)
+				case *ssa.Call:
+					result.calls = append(result.calls, v)
+				case *ssa.FieldAddr:
+					result.fieldAddr = append(result.fieldAddr, v)
+				case *ssa.Store:
+					result.store = append(result.store, v)
+				}
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func TestSource(t *testing.T) {
+	dir := analysistest.TestData()
+	config := &testConfig{
+		propagatorsPattern: "propagator",
+		sanitizerPattern:   "sanitizer",
+		fieldsPattern:      "name",
+	}
+
+	testCases := []struct {
+		desc                         string
+		pattern                      string
+		config                       *testConfig
+		wantCallConnectionIdx        []int
+		wantFieldAccessConnectionIdx []int
+		wantStoreConnectionIdx       []int
+	}{
+		{
+			desc:                   "allocation",
+			pattern:                "allocation",
+			config:                 config,
+			wantStoreConnectionIdx: []int{0},
+		},
+		{
+			desc:                         "propagation",
+			pattern:                      "propagation",
+			config:                       config,
+			wantCallConnectionIdx:        []int{0},
+			wantFieldAccessConnectionIdx: []int{0},
+		},
+		{
+			desc:                         "sanitization",
+			pattern:                      "sanitization",
+			config:                       config,
+			wantFieldAccessConnectionIdx: []int{0},
+		},
+	}
+
+	for _, tt := range testCases {
```

Comment:
This took me a while to understand, and I think every check done here would be performed by a `// want "allocation"`.

Reply:
These tests go beyond just identifying where allocation is performed for a Source.
Concretely, these tests ensure that we:
1. do not build the graph of a Source beyond a sanitizer
2. do include calls to propagators in the graph